### PR TITLE
Add testing utilities for service transcripts.

### PR DESCRIPTION
**Pull request by hdevalence**
This adds a `Transcript` type that wraps an iterator of request-response pairs.  The `Transcript` can do two things: it can `check` a service against the transcript, or it can work as a `Service` impl itself, allowing testing of how one service reacts to another.

This API is used to factor all of the common code out of the `zebra_state` tests.

### CODE REVIEW

File: zebra-test/src/transcript.rs
Comment (yaahc) [inline zebra-test/src/transcript.rs:48]: :heart_eyes:

File: zebra-state/tests/basic.rs
Comment (yaahc) [inline zebra-state/tests/basic.rs:52]: I think it might be a good idea to have individual tests for each transcript so they can be run independently

File: zebra-state/tests/basic.rs
Comment (yaahc) [inline zebra-state/tests/basic.rs]: Once this merges I'll fix the install_tracing stuff

**Review by yaahc — DISMISSED** [review]
sugoi

File: zebra-test/src/transcript.rs
Comment (hdevalence) [reply zebra-test/src/transcript.rs:48]: Do you know what the correct bound on `C::Error` to be able to turn the `unwrap()` into a `?` is? Ideally this should work on any `Service`.

File: zebra-state/tests/basic.rs
Comment (hdevalence) [reply zebra-state/tests/basic.rs:52]: I merged it all into one test to cut down on the common code so that it's very easy to extend the list of transcripts.  I guess the downside is that they can't be run independently, but because we're using `Report` we should get isolated errors on any failures, and I'm not sure how often we'd want to run individual transcripts?

File: zebra-state/tests/basic.rs
Comment (hdevalence) [reply zebra-state/tests/basic.rs]: No need to block on it, I'm going to be afk for the next few hours so if you prefer to fix the install_tracing stuff now, i'm happy to rebase this PR on top of yours and de-conflict things

File: zebra-test/src/transcript.rs
Comment (yaahc) [reply zebra-test/src/transcript.rs:48]: Aah, so this is where we get fucked by tower requiring that we return `Box` from buffers. This is related to needing to put `.map_err(|e| eyre!(e))` on every callback from a tower service. The `eyre` macro uses autoref specialization to handle the fact that `Box<dyn Error>` isnt an error and its impossible to implement `From<Box<dyn Error>> for Report`. This trick wont work in a generic context though, because we only have the universally quantified parameter, not the actual type we're working with. Ideally the bound would be `C::Error: Into<Report>`, but that doesn't work with `Box` which is the _only_ error type we actually use in practice .... 

_frustration_

So for now I'd say just require that the service is a `Box<dyn Error>` and put a `// TODO(jlusby): fix this once we fix tower error handling` so I remember to update it when we'll be able to actually change the bound to `Into<Report>`

File: zebra-state/tests/basic.rs
Comment (yaahc) [reply zebra-state/tests/basic.rs:52]: Once we have a lot of tests I expect if one starts failing we will want to be able to run `cargo test -- relevant_test_name` to do faster test and dev turn around, but if that comes up we can move the important transcript to its own test, there's no real reason to do it now.

**Pull request merged**
